File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var maxSubarrayLength = function(nums, k) {
    // Map to store the frequencies of elements
    const freq = new Map();

    let left = 0;
    let ans = 0;

    for (let right = 0; right < nums.length; right++) {
        const rNum = nums[right];
        freq.set(rNum, (freq.get(rNum) || 0) + 1);

        // Shrink the window from the left if the current element's frequency exceeds k
        while (freq.get(rNum) > k) {
            const lNum = nums[left];
            freq.set(lNum, freq.get(lNum) - 1);
            left++;
        }

        ans = Math.max(ans, right - left + 1);
    }

    return ans;
};

File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.py
class Solution:
    def maxSubarrayLength(self, nums: list[int], k: int) -> int:
        # Dictionary to store the frequencies of elements
        freq = {}

        left = 0
        ans = 0

        for right in range(len(nums)):
            r_num = nums[right]
            freq[r_num] = freq.get(r_num, 0) + 1

            # Shrink the window from the left if the current element's frequency exceeds k
            while freq[r_num] > k:
                l_num = nums[left]
                freq[l_num] -= 1
                left += 1

            ans = max(ans, right - left + 1)

        return ans
